API,MAINT: Reorganize array-wrap calling and introduce ``return_scalar``

seberg opened the pull request:
This reorganize how array-wrap is called.  It might very mildly change the semantics for reductions I think (and for negative priorities).

Overall, it now passes a new `return_scalar=False/True` when calling `__array_wrap__` and deprecates any array-wrap which does not accept `arr, context, return_scalar`.

I have not integrated it yet, but half the reason for the reorganization is to integrate it/reuse it in the `array_coverter` helper PR (gh-24214), which stumbled over trying to make the scalar handling sane.

Forcing downstream to add `return_scalar=False` to the signature is a bit annoying, but e.g. our memory maps currently try to _guess_ at it, which seems bad.  I am hoping, this can be part to making the scalar vs. array return more sane.

But, maybe mainly, I hope it consolidates things (together with gh-24124 mainly, as if ufuncs were the only complex place we used this, it wouldn't matter much).

seberg committed:
API: Reorganize `__array_wrap__` and add `return_scalar=False`

This also deprecates any `__array_wrap__` which does not accept
`context` and `return_scalar`.

seberg committed:
BUG: Fix niche bug in rounding.

seberg committed:
MAINT: Adjust __array_wrap__ in code and tests (also deprecation test)

seberg committed:
MAINT: Use/move the simplest C-wrapping also

seberg committed:
DOC: Update doc and add release note

seberg commented on numpy/_core/src/multiarray/calculation.c:
```diff
@@ -700,7 +700,7 @@ PyArray_Round(PyArrayObject *a, int decimals, PyArrayObject *out)
     Py_DECREF(f);
     Py_DECREF(out);
-    if (ret_int) {
+    if (ret_int && ret != NULL) {
```
Sorry, unrelated and a bit annoying to trigger.  But the deprecations triggered the bug...

seberg committed:
STY: Make linter happy in old tests

seberg committed:
MAINT: Silence GCC warning (value cannot really be used uninitialized)

seberg committed:
MAINT: Small style fixes

mhvk commented:
Hmm, seeing this I'm really not sure it is worth breaking every package that implemented `__array_wrap__` for something so relatively small... In the end, most classes could check whether `ndim==0` and then do something based on that.

Another idea, that does not break the signature (but might well break implementations), is to pass on to `__array_wrap__` what numpy would return, i.e., an array or a scalar.

But there also remains something to be said for just more generally using array scalars, and let that be the API break. It might just make code simpler rather than more complex...

p.s.  Should stress that I haven't thought enough about this! I should really look more at your `try-0d-preservation` branch...

seberg commented:
It would be nice not to require this, but...

> In the end, most classes could check whether ndim==0 and then do something based on that.

No because 0-D doesn't imply that NumPy should be returning a scalar, after all that is what annoys us all the time!  I would like to be able to decide that `np.add(1, 1)` should maybe a scalar, but `np.add(np.array(1), 1)` should be an array.  Or `arr1d.sum(axis)` vs. `arr1d.sum(-1)` maybe/probably.

It doesn't matter much which cases are which, so long there is no absolute uniform 0-d is always array/scalar...

> is to pass on to `__array_wrap__` what numpy would return, i.e., an array or a scalar.

Might work out mostly *until* `dtype=object` hits and all is lost, since you need at least `obj.view(my_subclass)` to work, and it cannot.  And it isn't even really reasonable to check for `isinstance(obj, ndarray)`...

Which was, why I thought I would just call `result[()]` instead, but that also doesn't work because we don't know that the result defines such indexing semantics.

So, I would love a different idea, but the only thing I have left is piggy-backing on `context` or try/except indefinitely.  But I really dislike try/except indefinitely, so that would leave adding one more try/except branch as the current one breaks subclasses that rely on `context` (although I hope beyond masked arrays few exist, and everyone who might want to use it moves to `__array_ufunc__`).

seberg committed:
BUG: Fix reference leak in ufunc array-wrapping

This probably doesn't fix the 32bit issue unfortunately, only the windows ones...

seberg committed:
BUG: Fix leak for result arrays in all ufunc calls

seberg commented:
FWIW, I have updated the `try-0d-preservation` branch with these (and the array-converter) changes.  There are some tests still failing, but the last commit shows that it is not impossible anymore to return scalars for scalar inputs and arrays otherwise.
(With the admitted caveat in the strange case where e.g. `dtype="i,i"` means that a tuple is considered a scalar rather, when it normally is not.)

mhvk commented:
> Which was, why I thought I would just call result[()] instead, but that also doesn't work because we don't know that the result defines such indexing semantics.

Maybe we should still go with this? I think it may cause less breakage than a new argument... 

Though perhaps I'm overly worried; not sure how many projects in fact have an `__array_wrap__` to start with...

seberg commented:
Let's say this, we currently have two possible paths:
1. Functions which always return an array
2. Functions which return a scalar if the result is 0-D.

And `__array_wrap__` cannot possibly match that, because even if it relies on `context` being set for the second, it will miss reductions.

So to some degree, I really want to add the ability to pick the two behaviors consciously for NumPy.  Now, you could just not tell `__array_wrap__` about it as it rarely needs to know in practice (most implementations are already happy to just ignore it).

> I think it may cause less breakage than a new argument...

Well, in the sense of requiring less code change?  Yes, for sure.  In the sense of braking things _disregarding_ the deprecation warning, not sure?  Since I think some things like pandas Series will run into `result[()]` not working.

> Though perhaps I'm overly worried; not sure how many projects in fact have an `__array_wrap__` to start with...

Right, and if I add another step which tries to pass `arr, context` when `arr, context, return_scalar` fails, then this is a `DeprecationWarning` and will not fail straight away (albeit add a bunch of overhead).

I don't have a strong opinion, it seems tedious, but OTOH, it seems bound to be annoying for someone to not be able to match NumPy arrays (e.g. `mmap` is such a case).

seberg commented:
@mhvk sorry, need to come back to this, I think it is pretty important (if just to push the followup of `array_coverter()`, which *is* important to fixup NEP 50 niche case bugs in the Pytho API).

We had discussed this in a meeting a while ago, and I don't think anyone had serious concerns with adding it.  But I do value your opinion more than most when it comes to `__array_function__`.  What I will say is that this is just a deprecation, so yes, everyone must update, but end-users wouldn't see it immediately (although it's slow).  It could be a `PendingDeprecationWarning`, but that just hides it even more with no gain I think.  I can will add a try for `__array_function__(arr, context)`, so that it is truly just a deprecation (even if it could go through 2 trys before succeeding).

To me, the new kwarg still just seems easier than any other polymorphism to try to guess this without it, and it is only some libraries that need to adapt.

mhvk commented on numpy/_core/src/umath/ufunc_object.c:
```diff
@@ -4440,47 +4186,68 @@ replace_with_wrapped_result_and_return(PyUFuncObject *ufunc,
+        PyObject *ret_i = npy_apply_wrap(
+                (PyObject *)result_arrays[out_i], original_out, wrap, wrap_type,
+                /* Always try to return a scalar right now: */
+                &context, PyArray_NDIM(result_arrays[out_i]) == 0, NPY_TRUE);
```
Here, the forcing is consistent with the previous behaviour. I think `reduce` always behaved differently...

mhvk commented on numpy/_core/src/umath/ufunc_object.c:
```diff
@@ -4440,47 +4186,68 @@ replace_with_wrapped_result_and_return(PyUFuncObject *ufunc,
         }
-        for (int i = 0; i < ufunc->nout; i++) {
-            PyTuple_SET_ITEM(result, i, retobj[i]);
+        /* When we are not returning a tuple, this is the result: */
```
Nice way to use the loop and be able to exit it.

mhvk commented on numpy/_core/tests/test_deprecations.py:
```diff
@@ -787,3 +787,25 @@ def _check_for_warning(func):
+            def __array_wrap__(self, arr, context=None):
+                return 'pass'
+
+        class Test2:
```
Can inherit from `Test1` to allow removing the `__array__` method.

mhvk commented on numpy/_core/src/multiarray/arraywrap.c:
```diff
@@ -0,0 +1,277 @@
+    if (res == NULL) {
+        goto finish;
+    }
+    if (DEPRECATE(
```
Add a comment when this was deprecated (possibly for our future selves!).

mhvk reviewed (commented):
@seberg - sorry to have forgotten about this! Looking with a fresh eye, I think your solution is for the best -- everything I can think of has the risk of subtle breakage.

So, now just a proper review of the code -- which looks great! One genuine comment about whether forcing wrapping for reductions is correct (looks like the old code did not). Though if forced wrapping for reductions is OK, then all the calls to `npy_apply_wrap` in fact used forced, so the argument could be removed. Indeed, that suggests that if we do not want to force wrapping in reductions, we could just move the `if` statement to the reduction instead (or are you using it in the follow-up PR?).

seberg committed:
Ensure we try passing context and address related smaller review comments

seberg committed:
Merge branch 'main' into clean-wrapping

seberg committed:
Ensure we don't try `context=None` and expand code comment

seberg committed:
Rely on return_scalar always being right (and style nit)

seberg committed:
Remove outdated comments as per review

seberg reviewed (commented):
Thanks for the review, sorry, I have not gotten back earlier.  I think I addressed the issues (the largest one maybe being to try just passing context).

Forcing wrapping in the reduction path seemed right to me, but that said I don't care about just keeping things as they were.  It doesn't _really_ belong here, except that I wanted to slowly try to standardize behavior more.

seberg committed:
Let's just undo force-wrap for now for reductions (its a change...)

seberg commented:
Nevermind, I siwtched to not do force-wrap which I think leaves the reduction result unchanged.  While I would like to align them, I need this in more for the other PR then delay thinking about it.

seberg committed:
ENH: Chain the original error when the deprecationwarning is raised

Doing this due to gh-25635 since it is super confusing with the
bad retrying...

seberg commented:
gh-25635 was a good reason to chain exceptions.  Because this was *always* terrible (really it got no worse, except now you have a chance of getting to the real error).  It shows nicely how that try/except can go wrong if it isn't very specific enough.
```
In [4]: np.sqrt(np.ma.masked)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File ~/forks/numpy/build-install/usr/lib/python3.10/site-packages/numpy/ma/core.py:6686, in MaskedConstant.__array_wrap__(self, obj, context, return_scalar)
   6685 def __array_wrap__(self, obj, context=None, return_scalar=False):
-> 6686     return self.view(MaskedArray).__array_wrap__(obj, context)

File ~/forks/numpy/build-install/usr/lib/python3.10/site-packages/numpy/ma/core.py:3131, in MaskedArray.__array_wrap__(self, obj, context, return_scalar)
   3129     fill_value = self.fill_value
-> 3131 np.copyto(result, fill_value, where=d)
   3133 # Update the mask

TypeError: Cannot cast scalar from dtype('float64') to dtype('bool') according to the rule 'safe'

The above exception was the direct cause of the following exception:

DeprecationWarning                        Traceback (most recent call last)
Cell In[4], line 1
----> 1 np.sqrt(np.ma.masked)

DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
```

mhvk commented on numpy/_core/src/multiarray/arraywrap.c:
```diff
@@ -0,0 +1,308 @@
+    for (int i = 0; i < nin; i++) {
+        PyObject *obj = inputs[i];
+        if (PyArray_CheckExact(obj)) {
+            if (wrap == NULL || priority < 0) {
```
Priority cannot be `< 0` here, right? Should the initialization above be at `-inf`? A benefit of that is that I think the test for `wrap == NULL` would not be needed any more.

mhvk replied:
I'd also write `priority < NPY_PRIORITY` and use that below for `priority =  NPY_PRIORITY` - somewhat easier to read than having to know the priority of arrays is 0.

seberg replied:
Priority can be `< 0`, this makes sense for example for `memmap` and is used there.  

I had initially thought about it, and decided it seemed dubious.  Maybe all it changes is that `-inf` really means that `__array_wrap__` should *never* be called, which would make sense (and probably nobody uses that anyway).

There is the odd code, that the old code just *skipped* ndarrays, which is worse, because negative wraps take precendence over ndarray!  (yes this is a change here)

seberg replied:
(Will replace the `NPY_PRIORITY` to not hardcode the 0)

mhvk replied:
But if you initialize `priority = 0`, doesn't that mean that `np.memmap.__array_wrap__` never gets selected? Is that OK?

seberg replied:
The code here always uses the first one (if there is more than one).  That is what the `NULL` check does.

seberg replied:
Ah, that is nicer though of course!  I should write `i == 0` that is much clearer...

seberg replied:
Wait no, arg... that doesn't work, how about just adding a comment for what the `NULL` check does early on?

mhvk commented on numpy/_core/src/multiarray/arraywrap.c:
```diff
@@ -0,0 +1,308 @@
+            }
+        }
+        else if (PyArray_IsAnyScalar(obj)) {
+            if (wrap == NULL || priority < NPY_SCALAR_PRIORITY) {
```
Same here.

mhvk commented on numpy/_core/src/multiarray/arraywrap.c:
```diff
@@ -0,0 +1,308 @@
+            }
+        }
+        else {
+            PyObject *new_wrap = PyArray_LookupSpecial_OnInstance(obj, npy_ma_str_array_wrap);
```
I guess this can be inside the `if` statement below, since `new_wrap` is not used if the priority is too low. Though that means two lookups if `__array_wrap__` does not exist, so maybe what you have is better. If you keep it, though, I think you need to `Py_DECREF new_wrap` in an `else` clause (as it is a new reference).

seberg replied:
Ouch, thanks!  I agree with moving looking nicer, but then decided that maybe `__array_priority__` without `__array_wrap__` is actually not that uncommon, so just added the decref.

mhvk commented on numpy/_core/src/multiarray/arraywrap.c:
```diff
@@ -0,0 +1,308 @@
+        return_scalar = NPY_FALSE;
+
+        if (PyArray_CheckExact(original_out)) {
+            wrap = Py_None;
```
I stared at this for a long time, and thought that surely this needed `Py_INCREF(Py_None)`, but then I guess this is just a signal, and you leave `DECREF`ing wrap to the caller (logically), so this is fine. Maybe add a comment `/* Used as signal; wrap is not DECREF'd, so no need to INCREF(Py_None) here. */`

mhvk reviewed (commented):
Mostly small comments about the logic in `npy_find_array_wrap` (and one ref-counting issue).

seberg committed:
BUG,MAINT: Address Martens review comments

mhvk commented on numpy/_core/src/multiarray/arraywrap.c:
```diff
@@ -0,0 +1,318 @@
+    for (int i = 0; i < nin; i++) {
+        PyObject *obj = inputs[i];
+        if (PyArray_CheckExact(obj)) {
+            if (wrap == NULL || priority < NPY_PRIORITY) {
```
I'm still confused: if you initialize `priority = -inf` (or some other very large negative number), will this line not just always be true if `wrap` is undefined? So, why is the `wrap == NULL` needed?

seberg replied:
not if the object defines `priority = -inf`.  Maybe that would be preferable, but :shrug:.

seberg replied:
Ah, I only need it on the last branch, yes.  Would that seem clearer (then the last branch is still special, this way priority is only initialized to silence warnings)?

mhvk replied:
I think coding against something that defines `priority = -inf` yet defined `__array_wrap__` seems unnecessary! But if you really prefer what you have, that is fine too -- my own sense is that the logic becomes clearer without the extra conditions that are needed to grab the first argument unconditionally.

seberg replied:
I don't disagree, I think I just wanted to match *one* of the implementations inside NumPy.
Maybe it is laziness speaking: This way no user can run into this (not even hypothesis) and its there already there...?

mhvk reviewed (approved):
OK, I think we've reached (more than) good enough!

You probably want to squash & merge, or reduce the number of commits a bit.

seberg commented:
Thanks Marten, let me squash merge this then, since I want to rebase the *other* PR on this work...

Afterall, while I think it fixes a bunch of issues (at least mid-term) and opens avenues, it was really mostly a split-out from the other PR.

mhvk commented:
Nice, on to the helper! (ping me when rebased - that one we went over a lot already so should be easy to do a final round of review).

pllim commented:
I thought numpy 2 would be out by now. Wasn't expecting breaking changes even now...

seberg commented:
> I thought numpy 2 would be out by now. Wasn't expecting breaking changes even now...

I wish :(, although I hope it should be mostly good.  You will certainly also notice gh-25168 (more than this).
